### PR TITLE
feat/message filter

### PR DESCRIPTION
**Pull request by Mirko-von-Leipzig**
Adds the `MessageFilter` and the `unsolicitation_listener` test case.

I added quite a bit of docs to the `MessageFilter` - mostly as an exercise to learn Rust docs.

### CODE REVIEW

File: src/protocol/message.rs [hunk 1]
```diff
@@ -66,6 +66,7 @@ impl MessageHeader {
     }
 }
 
+#[derive(Debug)]
```
Comment (niklaslong) [inline src/protocol/message.rs:69]: Thanks for adding these! 👍

File: src/tests/conformance/messages.rs [hunk 2]
```diff
@@ -51,3 +51,32 @@ async fn handshake(node_addr: SocketAddr) -> io::Result<TcpStream> {
 
     Ok(peer_stream)
 }
+
+#[tokio::test]
+async fn unsolicitation_listener() {
```
Comment (niklaslong) [inline src/tests/conformance/messages.rs:56]: As this test isn't part of the index, we may want to ignore it by default (with `#[ignore]`)?

File: src/protocol/message.rs [hunk 3]
```diff
@@ -179,3 +180,187 @@ fn checksum(bytes: &[u8]) -> u32 {
...
+            Message::GetHeaders(_) => {
+                Message::Headers(Headers::empty())
+                    .write_to_stream(stream)
+                    .await
+            }
+            Message::GetData(inv) => Message::NotFound(inv).write_to_stream(stream).await,
+            _ => unimplemented!(),
+        };
+
+        result
```
Comment (niklaslong) [inline src/protocol/message.rs:308]: Suggestion: instead assigning the result, you could use `?` to bubble the error and return `Ok(())` if the execution isn't short-circuited.

File: src/protocol/message.rs [hunk 4]
```diff
@@ -179,3 +180,187 @@ fn checksum(bytes: &[u8]) -> u32 {
 
     u32::from_le_bytes(checksum)
 }
+
+/// Controls the filter response of [MessageFilter] to messages it reads.
+#[derive(Debug, Clone, Copy, PartialEq)]
+pub enum Filter {
```
Comment (niklaslong) [inline src/protocol/message.rs:186]: Future consideration: we may want to move this code to another directory as it isn't explicitely part ot the protocol. Maybe `src/setup/message_filtering.rs` or similar? We'll also have some handshake helpers there and in future the synthetic nodes.

File: src/protocol/message.rs [hunk 5]
```diff
@@ -179,3 +180,187 @@ fn checksum(bytes: &[u8]) -> u32 {
...
+    pub fn with_ping_filter(mut self, filter: Filter) -> Self {
+        self.ping = filter;
+        self
+    }
+
+    // sends an appropriate reply in response to the received message
+    async fn reply(&self, stream: &mut TcpStream, message: Message) -> io::Result<()> {
+        let result = match message {
+            Message::Ping(nonce) => Message::Pong(nonce).write_to_stream(stream).await,
+            Message::GetAddr => Message::Addr(Addr::empty()).write_to_stream(stream).await,
```
Comment (niklaslong) [inline src/protocol/message.rs:298]: I think this is fine for now but we'll want more flexibility in future (e.g. returning the addresses a synthetic node is connected to in the mesh for load testing).

File: src/protocol/payload/block.rs [hunk 6]
```diff
@@ -46,12 +47,20 @@ impl LocatorHashes {
...
 }
 
+#[derive(Debug)]
 pub struct Headers {
     count: VarInt,
     headers: Vec<Header>,
 }
 
 impl Headers {
+    pub fn empty() -> Self {
```
Comment (niklaslong) [inline src/protocol/payload/block.rs:57]: I like how explicit this is, we may want to add `Default` implementations in future if this is a pattern we find ourselves using a lot. Just something to keep in mind.

**Review by niklaslong — COMMENTED** [review]
This looks great, I left some suggestions and a few more general comments/things to consider in future. 

The documentation is also great, thanks for being so thorough! 🚀

File: src/protocol/message.rs (hunk 4 above)
Comment (Mirko-von-Leipzig) [reply src/protocol/message.rs:186]: Yes, I was quite uncertain about where to place it. Considered `src/setup/message_filtering.rs`, or creating a more utility focused module. Or even a module for generating fake messages which we may need in the future.

For now I just wanted to keep it compact while we figure it out organically.

File: src/protocol/message.rs (hunk 3 above)
Comment (Mirko-von-Leipzig) [reply src/protocol/message.rs:308]: That's what I get for copy pasting a previous attempt 🤦

I can in-fact simply return the result straight from the match.

**Pull request merged**